### ISSUE TITLE
Double tf.log()'s when calculating CategoricalPd.sample() ?

### ISSUE THREAD

**Issue by keven425** [openai/baselines#172]
https://github.com/openai/baselines/blob/4993286230ac92ead39a66005b7042b56b8598b0/baselines/common/distributions.py#L155

On this line above, it takes log twice on u. Why is that the case? I thought it should only take log once, like this:

```
return tf.argmax(self.logits -tf.log(u), axis=-1)
```

@joschu

**Comment by unixpickle** [comment]
`-log(-log(uniform))` is the Gumbel distribution. This use a Gumbel-Softmax to sample from a uniform distribution.
